### DISCUSSION TITLE
Custom Controllers?

### DISCUSSION THREAD

**Discussion by carolynvs** [fluxcd/flux2#676] [Q&A] [answered]
I would like to integrate [Porter](https://porter.sh) with Flux. This allows people to use Flux for more complex application deployments that don't fit neatly into just helm charts. 

The [source watcher example](https://toolkit.fluxcd.io/dev-guides/source-watcher/) is very helpful, thanks! I am wondering if you have a recommendation for people developing custom controllers? Are you interested in having them contribute mature controllers back upstream, and add commands to support them to fluxctl?

For example, fluxctl doesn't appear to support third parties adding commands for custom controllers through something like plugins. So I am going to either fork fluxctl or just write a custom cli. First time setup will involve installing flux, my custom flux controller, the underlying operator and a custom cli.

And that's not a problem! But I did want to understand what your vision is for custom controllers, in case I am making assumptions that you would prefer new controllers be maintained outside of Flux, through forks.

Thanks!

**Comment by stefanprodan** [comment, marked as answer]
We plan to have a Flux components catalog for external controllers that subscribe to the GitOps Toolkit conventions by using [fluxcd/pkg/runtime](https://pkg.go.dev/github.com/fluxcd/pkg/runtime) and [fluxcd/pkg/apis/meta](https://pkg.go.dev/github.com/fluxcd/pkg/apis/meta) packages. As for extending Flux CLI, this could be done in the same fashion as git or kubectl, by placing a binary `flux-<controller-name>` in the path.

**Comment by stefanprodan** [comment]
> Are you interested in having them contribute mature controllers back upstream, and add commands to support them to fluxctl?

Yes! We need to define what "mature" means for us, but a specialised reconciler for Porter would be a great candidate being a CNCF project.

**Comment by carolynvs** [comment]
I'm so glad that I asked! A fluxctl plugin will work perfectly.

Thanks again for working on Flux! 🙇‍♀️
